includes/subscriptions/class-wc-payments-invoice-service.php

```diff
@@ -116,41 +116,24 @@ public function maybe_record_first_invoice_payment( int $order_id, string $old_s
 	 * @param string          $customer_id            The WCPay Customer ID.
 	 * @param string          $wcpay_subscription_id  The WCPay Billing subscription ID.
 	 *
-	 * @throws \Exception When there's an error preparing the invoice item data.
+	 * @throws API_Exception When there's an error creating the invoice items on server.
 	 *
-	 * @return string[]|WP_Error Invoice item ids.
+	 * @return string[] Invoice item ids.
 	 */
 	public function create_invoice_items_for_subscription( WC_Subscription $subscription, string $customer_id, string $wcpay_subscription_id = '' ) {
-		$invoice_item_ids = [];
+		$invoice_item_data = $this->prepare_invoice_item_data( $subscription );
+		$request_data      = [];
 
-		try {
-			$invoice_item_data = $this->prepare_invoice_item_data( $subscription );
-
-			if ( is_wp_error( $invoice_item_data ) ) {
-				throw new \Exception( 'There was a problem preparing invoice item data: ' . wp_json_encode( $invoice_item_data ) );
+		foreach ( $invoice_item_data as $data ) {
+			if ( $wcpay_subscription_id ) {
+				$data['subscription'] = $wcpay_subscription_id;
 			}
 
-			foreach ( $invoice_item_data as $data ) {
-				if ( $wcpay_subscription_id ) {
-					$data['subscription'] = $wcpay_subscription_id;
-				}
-
-				$data['customer'] = $customer_id;
-				$result           = $this->payments_api_client->create_invoice_item( $data );
-
-				if ( is_wp_error( $result ) ) {
-					throw new \Exception( "There was a problem creating the {$data['description']} invoice item: " . wp_json_encode( $result ) );
-				}
-
-				$invoice_item_ids[] = $result['id'];
-			}
-		} catch ( \Exception $e ) {
-			$this->delete_invoice_items( $invoice_item_ids );
-
-			return new WP_Error( $e->getMessage() );
+			$data['customer'] = $customer_id;
```

Comment:
What do you think about handling `subscription` and `customer` params in `prepare_invoice_item_data`? We could pass in both IDs and set them there instead?

And in the case of `$wcpay_subscription_id`, we might even be able to just use `subscription` which is already an argument for `prepare_invoice_item_data` via `WC_Payments_Subscription_Service::get_wcpay_subscription_id`

Reply:
I like this idea! I'ved updated the `prepare_invoice_item_data()` to now populate the customer/subscription ID fields: d25181f

> And in the case of $wcpay_subscription_id, we might even be able to just use subscription which is already an argument for prepare_invoice_item_data via WC_Payments_Subscription_Service::get_wcpay_subscription_id

I stuck with keeping the `$wcpay_subscription_id` param being passed to this function even though we can fetch it from the subscription to avoid double fetching this data (since it's also needed in those functions where we call `create_invoice_items_for_subscription()`

If we ever do call `create_invoice_items_for_subscription()` where we haven't already fetched the WCPay Subscription ID, we could add some handling to fetch it from the subscription in `prepare_invoice_item_data()` but I didn't see a need for this right now, so I've left it out.

Let me know if that sits alright with you 😄

Reply:
That's fine with me.  Thank you!